Userland/Libraries/LibCrypto/Hash/SHA1.cpp

```diff
@@ -67,6 +83,188 @@ inline void SHA1::transform(u8 const* data)
+    e = _mm_sha1nexte_epu32(abcd_1, msg_1);
+    abcd_1 = _mm_sha1rnds4_epu32(abcd_2, e, 2);
+    // Generate next input, aka preprocessing, described in NIST FIPS PUB 180-4 section 6.1.2.1.
+    msg_2 = _mm_sha1msg1_epu32(msg_2, msg_3);
+    msg_2 = _mm_xor_si128(msg_2, msg_0);
+    msg_2 = _mm_sha1msg2_epu32(msg_2, msg_1);
+    // Four rounds (56-59) of the SHA-1 algorithm.
+    e = _mm_sha1nexte_epu32(abcd_2, msg_2);
+    abcd_2 = _mm_sha1rnds4_epu32(abcd_1, e, 2);
+    // Generate next input, aka preprocessing, described in NIST FIPS PUB 180-4 section 6.1.2.1.
+    msg_3 = _mm_sha1msg1_epu32(msg_3, msg_0);
+    msg_3 = _mm_xor_si128(msg_3, msg_1);
+    msg_3 = _mm_sha1msg2_epu32(msg_3, msg_2);
+    // Four rounds (60-63) of the SHA-1 algorithm.
+    e = _mm_sha1nexte_epu32(abcd_1, msg_3);
+    abcd_1 = _mm_sha1rnds4_epu32(abcd_2, e, 3);
+    // Generate next input, aka preprocessing, described in NIST FIPS PUB 180-4 section 6.1.2.1.
+    msg_0 = _mm_sha1msg1_epu32(msg_0, msg_1);
+    msg_0 = _mm_xor_si128(msg_0, msg_2);
+    msg_0 = _mm_sha1msg2_epu32(msg_0, msg_3);
+    // Four rounds (64-67) of the SHA-1 algorithm.
+    e = _mm_sha1nexte_epu32(abcd_2, msg_0);
+    abcd_2 = _mm_sha1rnds4_epu32(abcd_1, e, 3);
+    // Generate next input, aka preprocessing, described in NIST FIPS PUB 180-4 section 6.1.2.1.
+    msg_1 = _mm_sha1msg1_epu32(msg_1, msg_2);
+    msg_1 = _mm_xor_si128(msg_1, msg_3);
+    msg_1 = _mm_sha1msg2_epu32(msg_1, msg_0);
+    // Four rounds (68-71) of the SHA-1 algorithm.
+    e = _mm_sha1nexte_epu32(abcd_1, msg_1);
+    abcd_1 = _mm_sha1rnds4_epu32(abcd_2, e, 3);
+    // Generate next input, aka preprocessing, described in NIST FIPS PUB 180-4 section 6.1.2.1.
+    msg_2 = _mm_sha1msg1_epu32(msg_2, msg_3);
+    msg_2 = _mm_xor_si128(msg_2, msg_0);
+    msg_2 = _mm_sha1msg2_epu32(msg_2, msg_1);
+    // Four rounds (72-77) of the SHA-1 algorithm.
+    e = _mm_sha1nexte_epu32(abcd_2, msg_2);
+    abcd_2 = _mm_sha1rnds4_epu32(abcd_1, e, 3);
+    // Generate next input, aka preprocessing, described in NIST FIPS PUB 180-4 section 6.1.2.1.
+    msg_3 = _mm_sha1msg1_epu32(msg_3, msg_0);
+    msg_3 = _mm_xor_si128(msg_3, msg_1);
+    msg_3 = _mm_sha1msg2_epu32(msg_3, msg_2);
+    // Four rounds (76-79) of the SHA-1 algorithm.
+    e = _mm_sha1nexte_epu32(abcd_1, msg_3);
+    abcd_1 = _mm_sha1rnds4_epu32(abcd_2, e, 3);
+    // Finish computation of the last `e' value.
+    msg_0 = _mm_setzero_si128();
+    e = _mm_sha1nexte_epu32(abcd_2, msg_0);
+    // Sum current working registers with saved state from before.
+    abcd_1 = _mm_add_epi32(abcd_1, old_abcd);
+    e = _mm_add_epi32(e, old_e);
+    // Here could end `for' or `while' loop in case we processed more than one block at once.
+    // Save working registers into state.
+    abcd_1 = _mm_shuffle_epi32(abcd_1, 0x1b);
+    _mm_storeu_si128(reinterpret_cast<__m128i*>(&state[0]), abcd_1);
+    *reinterpret_cast<int*>(&state[4]) = _mm_extract_epi32(e, 3);
```

Comment:
This is cast and storage is not safe, for alignment reasons.

Can we mark alignment requirements on the input state variable? If it's a member variable of the SHAXX class, we should absolutely set an alignment requirement on it, to avoid hard to debug trap/crashes.

We have AK::ByteReader for this.

Reply:
No, it is safe. The write goes to variable of type `u32` and the read is a "cast" from (part of) 128bit register. I could improve this by `static_cast`ing the `int` to `u32`.

Reply:
Added `static_cast`.